### ISSUE TITLE
[22.06-beta] Containerd binaries are missing from Windows package

### ISSUE THREAD

**Issue by olljanat** [moby/moby#43892]
### Description

In my opinion [Windows static binaries](https://download.docker.com/win/static/test/x86_64/docker-22.06.0-beta.0.zip) package should contains containerd and its add-ons like [Linux package](https://download.docker.com/linux/static/test/x86_64/docker-22.06.0-beta.0.tgz) does in 22.06.x versions and documentation should explain how to switch to that runtime with warning that it is first implementation and might contain bugs.

Preferably containerd should be installed as service but leaved to disabled state by default which means that [install-docker-ce.ps1](https://github.com/microsoft/Windows-Containers/blob/Main/helpful_tools/Install-DockerCE/install-docker-ce.ps1) which [this](https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/set-up-environment?tabs=dockerce#windows-server-1) document is refering should be updated to do so.

However I'm not aware of about any real showstopper bugs but instead of those are more likely related to challenges to do proper CI testing for all features and there is some small corner cases.

### Reproduce

-

### docker version

```bash
-
```

### docker info

```bash
-
```

**Comment by olljanat** [comment]
@thaJeztah @TBBle your thoughts?

**Comment by TBBle** [comment]
It does make sense to me for the static binaries zip to include dependencies, particularly if the plan is still to get containerd as the default backend for Windows Server 2022 onwards. A specific expectation/benefit for me of this is that the zip includes the exact containerd version which was used in CI, reducing the scope for mysterious bugs.

**Comment by olljanat** [comment]
> A specific expectation/benefit for me of this is that the zip includes the exact containerd version which was used in CI, reducing the scope for mysterious bugs.

Yes, this was my thinking also which why it would be better to provide containerd binaries with Docker and to be in line with Linux version.

> Given Docker Desktop's role in the ecosystem, it would make sense to have its own docker-containerd service, but I don't feel that's _as true_ for the various other dockerd.exe installation options.

There looks to be actually two different ways how this is handled in Linux. When installing to with apt or rpm containerd gets installed as separate service as those tools are able to require specific version. But when installing with static binaries there is no containerd service at all, only docker service and that will start containerd as process which it manages https://github.com/moby/moby/blob/90fce781d9a16f367a87d2784b7987ec7fecf7a9/cmd/dockerd/daemon_unix.go#L140-L151
and in that case container is configured to listening `/var/run/docker/containerd/containerd.sock` instead of default one.

So probably best option would be to use same method in Windows too or we can go even further and always run separate containerd process even if there is containerd service available. Then switching to container should be as simple as add `{"default-runtime": "containerd"}` to daemon config.
 
> particularly if the plan is still to get containerd as the default backend for Windows Server 2022 onwards. 

That was my hope but I don't think that it is realistic as there is no enough contributors who care about Windows in here.

**Comment by thaJeztah** [comment]
We're planning to split the static packages into separate downloads for each component; currently we bundle containerd and runc as part of a single tar / zip, which complicates releases (e.g. if there would be a new containerd release we can release new `.deb` and `.rpm` packages, but can't ship new static / binary releases unless we also do a new release of the engine); same for (e.g.) buildx or compose.

I agree that (now that using containerd as runtime for windows is an option) it'd be good to also publish packages for them; no decision was made yet on wether or not it should be the "default" configuration, but at least having it as a (easier to consume) option makes sense.

**Comment by olljanat** [comment]
> We're planning to split the static packages into separate downloads for each component; currently we bundle containerd and runc as part of a single tar / zip, which complicates releases (e.g. if there would be a new containerd release we can release new `.deb` and `.rpm` packages, but can't ship new static / binary releases unless we also do a new release of the engine); same for (e.g.) buildx or compose.

Old [DockerMsftProvider](https://github.com/OneGet/MicrosoftDockerProvider) PowerShell module used metadata file from [here](https://go.microsoft.com/fwlink/?LinkID=825636&clcid=0x409). One way to solve it is of course would be publish similar metadata which just connects Docker binaries to proper version of containerd binaries.

Then it can be handled in [install-docker-ce.ps1](https://github.com/microsoft/Windows-Containers/blob/Main/helpful_tools/Install-DockerCE/install-docker-ce.ps1) like I originally proposed.


PS. There looks to be also quite ugly bug on that one https://github.com/microsoft/Windows-Containers/issues/256 which you should be aware of if you see weird Docker versions in issues which people are reporting.

**Comment by crazy-max** [comment]
I have added build of containerd for windows in https://github.com/moby/moby/pull/43529.

```console
$ docker buildx bake all --set *.platform=windows/amd64
$ tree -nh ./bundles/all/
[4.0K]  ./bundles/all/
├── [ 403]  checksums.txt
├── [ 35M]  containerd.exe
├── [103K]  containerutility.exe
├── [ 17M]  ctr.exe
├── [1.9M]  docker-proxy.exe
└── [ 55M]  dockerd.exe

0 directories, 6 files
```

`containerd-shim-runhcs-v1` is not included, not sure we would need it for testing.

**Comment by olljanat** [comment]
> I have added build of containerd for windows in #43529, see [eaa0fa1](https://github.com/moby/moby/commit/eaa0fa155a58c1d8d2192c7e8f72f879cab1e541)

I'm not sure if make sense to rebuild containerd binaries here. Dockerfile.windows already contains logic to download those and GitHub workflow copies them out from container.

> `containerd-shim-runhcs-v1` is not included, not sure we would need it for testing.

It is kind of important like explained on https://container42.com/2022/01/10/shim-shiminey-shim-shiminey/

**Comment by crazy-max** [comment]
containerd cross comp itself with buildx/Buildkit for Windows without using a Windows image: https://github.com/containerd/containerd/runs/7730870736?check_suite_focus=true#step:6:10

Dockerfile: https://github.com/containerd/containerd/blob/main/.github/workflows/release/Dockerfile

**Comment by TBBle** [comment]
> I think the plan is to cross comp in the future and remove the need to use the `Dockerfile.windows`.

Removing `Dockerfile.windows` and fully relying on BuildKit would shut out building on Windows platforms where running Linux Docker is not feasible due to e.g., lack of nested virtualisation support. GitHub Actions runners fall into this category. VMs from a cloud provider my employer uses also have this limitation.

The only upstream-supported mechanism (AFAIK) for running Linux Docker on Windows Server is to directly install it under a WSL2 distribution, and that is only present on Windows Server 2022 (and still requires nested virtualisation support).

**Comment by olljanat** [comment]
> I think the plan is to cross comp in the future and remove the need to use the `Dockerfile.windows`. But needs first #43529 so we can just use the same workflow for win/linux to build binaries for testing. 

Ok but I still don't understand that why you want re-build containerd binaries for Windows in here instead of just down them from containerd GitHub? There is no multiple different distros of Windows where there would be different kind of static/dynamic binaries needed so IMO it is just wante of CI resources.

> Removing `Dockerfile.windows` and fully relying on cross-compilation would shut out building on Windows platforms where running Linux Docker is not feasible due to e.g., lack of nested virtualisation support. GitHub Actions runners fall into this category. 

GitHub Actions workflow supports sharing artifacts between runners so you can build in Windows binaries in Ubuntu and test in Windows. That works fine.


For me `Dockerfile.windows` removing is OK and probably even good idea as it have been quite ugly workaround. However we should improve documentation how to develop Windows version of Docker with Windows. I did share something in https://github.com/moby/moby/issues/43304#issuecomment-1193355184 as I noticed that it is now  possible build Docker that way which does not work in 20.10 branch.

**Comment by olljanat** [comment]
Now when #47955 was merged it would make even more sense to have containerd binaries included to those apckages.

**Comment by olljanat** [comment]
Nowadays with https://gist.github.com/profnandaa/9ad40dbd90ebad99896ee0fed2916406 it is also possible to use BuildKit in Windows so it would be nice to get it and buildx also included to Windows packages.

**Comment by vvoland** [comment]
We (Docker Inc., as Moby itself does not distribute any packages) are in progress of modernizing our build/release pipeline. Once we do that we will be able to build and distribute Windows containerd binaries, so that's definitely on our mind 👍🏻
